plugins/notifications-backend/src/service/DefaultNotificationRecipientResolver.ts

```diff
@@ -0,0 +1,170 @@
+import { NotificationRecipientResolver } from '@backstage/plugin-notifications-node';
+
+const isUserEntityRef = (ref: string) =>
+  parseEntityRef(ref).kind.toLocaleLowerCase() === 'user';
+
+// Partitions array of entity references to two arrays; user entity refs and other entity refs
+const partitionEntityRefs = (refs: string[]): string[][] =>
+  refs.reduce(
+    ([userEntityRefs, otherEntityRefs]: string[][], ref: string) => {
+      return isUserEntityRef(ref)
+        ? [[...userEntityRefs, ref], otherEntityRefs]
+        : [userEntityRefs, [...otherEntityRefs, ref]];
+    },
+    [[], []],
+  );
+
+export class DefaultNotificationRecipientResolver
+  implements NotificationRecipientResolver
+{
+  constructor(
+    private readonly auth: AuthService,
+    private readonly catalog: CatalogService,
+  ) {}
+
+  async resolveNotificationRecipients(options: {
+    entityRef: string | string[] | null;
+    excludeEntityRefs?: string | string[];
+  }): Promise<{ userEntityRefs: string[] }> {
+    const { entityRef, excludeEntityRefs = [] } = options;
+
+    if (entityRef === null) {
+      return { userEntityRefs: [] };
+    }
+
+    const excluded = [excludeEntityRefs].flat();
+    const refsArr = [entityRef].flat();
+
+    const [userEntityRefs, otherEntityRefs] = partitionEntityRefs(refsArr);
+    const users: string[] = userEntityRefs.filter(
+      ref => !excluded.includes(ref),
+    );
+    const entityRefs = otherEntityRefs.filter(ref => !excluded.includes(ref));
+
+    const fields = ['kind', 'metadata.name', 'metadata.namespace', 'relations'];
+    let entities: Array<Entity | undefined> = [];
+    if (entityRefs.length > 0) {
+      const fetchedEntities = await this.catalog.getEntitiesByRefs(
+        {
+          entityRefs,
+          fields,
+        },
+        { credentials: await this.auth.getOwnServiceCredentials() },
+      );
+      entities = fetchedEntities.items;
+    }
+
+    const mapEntity = async (entity: Entity | undefined): Promise<string[]> => {
+      if (!entity) {
+        return [];
+      }
+
+      const currentEntityRef = stringifyEntityRef(entity);
+      if (excluded.includes(currentEntityRef)) {
+        return [];
+      }
+
+      if (isUserEntity(entity)) {
+        return [currentEntityRef];
+      }
+
+      if (isGroupEntity(entity)) {
+        if (!entity.relations?.length) {
+          return [];
+        }
+
+        const groupUsers = entity.relations
+          .filter(
+            relation =>
+              relation.type === RELATION_HAS_MEMBER &&
+              isUserEntityRef(relation.targetRef),
+          )
+          .map(r => r.targetRef);
+
+        const childGroupRefs = entity.relations
+          .filter(relation => relation.type === RELATION_PARENT_OF)
+          .map(r => r.targetRef);
+
+        let childGroupUsers: string[][] = [];
+        if (childGroupRefs.length > 0) {
+          const childGroups = await this.catalog.getEntitiesByRefs(
```

Comment:
Hm. Is this worth memoising per ref? Will it be common to re-fetch groups over and over again?